File: leetcode/1210_Minimum_Moves_to_Reach_Target_with_Rotations.py

from unittest import TestCase
from typing import List, Tuple, Iterator, Set
from collections import deque

class Solution:
    def minimumMoves(self, grid: List[List[int]]) -> int:
        def succcessors(state: Tuple[int, int, bool]) -> Iterator[Tuple[int, int, bool]]:
            x, y, horizontal = state
            if horizontal:
                if x + 1 < n and y + 1 < n and grid[x + 1][y] == 0 and grid[x + 1][y + 1] == 0:
                    yield (x + 1, y, True)
                if x + 1 < n and y + 1 < n and grid[x + 1][y] == 0 and grid[x + 1][y + 1] == 0:
                    yield (x, y, False)
                if y + 2 < n and grid[x][y + 2] == 0:
                    yield (x, y + 1, True)
            else:
                if x + 1 < n and y + 1 < n and grid[x][y + 1] == 0 and grid[x + 1][y + 1] == 0:
                    yield (x, y + 1, False)
                if y + 1 < n and x + 1 < n and grid[x][y + 1] == 0 and grid[x + 1][y + 1] == 0:
                    yield (x, y, True)
                if x + 2 < n and grid[x + 2][y] == 0:
                    yield (x + 1, y, False)

        n = len(grid)
        state = (0, 0, True)
        queue = deque([(0, state)])
        visited: Set[Tuple[int, int, bool]] = { state }
        while queue:
            cost, state  = queue.popleft()
            if state == (n - 1, n - 2, True):
                return cost
            for next_state in succcessors(state):
                if next_state not in visited:
                    visited.add(next_state)
                    queue.append((cost + 1, next_state))
        return -1

class MinimumMovesTest(TestCase):
    def test_1(self) -> None:
        grid = [
            [0,0,0,0,0,1],
            [1,1,0,0,1,0],
            [0,0,0,0,1,1],
            [0,0,1,0,1,0],
            [0,1,1,0,0,0],
            [0,1,1,0,0,0]]
        self.assertEqual(11, Solution().minimumMoves(grid))

    def test_2(self) -> None:
        grid = [[0,0,1,1,1,1],
               [0,0,0,0,1,1],
               [1,1,0,0,0,1],
               [1,1,1,0,0,1],
               [1,1,1,0,0,1],
               [1,1,1,0,0,0]]
        self.assertEqual(9, Solution().minimumMoves(grid))

    def test_3(self) -> None:
        grid = [
            [0,0,1,0,0,0,0,0,0,0,0,0,0,0,0],
            [0,1,0,1,1,0,0,1,0,0,0,0,1,0,0],
            [0,1,0,0,0,0,1,0,0,1,0,0,0,0,0],
            [0,0,0,0,0,0,1,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,1,1,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0,1,0,1,0,0,0],
            [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
            [0,0,0,1,0,1,0,0,1,0,0,0,1,0,0],
            [0,0,0,0,1,0,0,0,0,0,0,0,0,1,0],
            [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],
            [0,0,0,1,0,0,0,0,0,0,0,0,0,0,0],
            [1,0,1,1,0,0,0,0,0,0,0,0,0,0,0],
            [0,0,0,0,0,0,0,0,0,0,0,0,0,1,0],
            [1,0,0,0,0,0,1,0,0,0,1,0,0,0,1],
            [0,0,1,0,1,0,0,0,0,0,0,0,0,0,0]]
        self.assertEqual(-1, Solution().minimumMoves(grid))

File: leetcode/1210_Minimum_Moves_to_Reach_Target_with_Rotations.rs
use std::collections::{HashSet, VecDeque};

struct Solution {}

impl Solution {
    pub fn minimum_moves(grid: Vec<Vec<i32>>) -> i32 {
        fn succesors(grid: &Vec<Vec<i32>>, n: usize, state: (usize, usize, bool)) -> Vec<(usize, usize, bool)> {
            let mut res = vec![];
            let (x, y, horisontal) = state;
            if horisontal {
                if y + 2 < n && grid[x][y + 2] == 0 {
                    res.push((x , y + 1, true));
                }
                if x + 1 < n && y + 1 < n && grid[x + 1][y] == 0 && grid[x + 1][y + 1] == 0 {
                    res.push((x + 1, y, true));
                }
                if x + 1 < n && y + 1  < n && grid[x + 1][y + 1] == 0 && grid[x + 1][y] == 0 {
                    res.push((x, y, false));
                }
            } else {
                if x + 2 < n && grid[x + 2][y] == 0 {
                    res.push((x + 1, y, false));
                }
                if x + 1 < n && y + 1 < n && grid[x][y + 1] == 0 && grid[x + 1][y + 1] == 0 {
                    res.push((x , y + 1, false));
                }
                if x + 1 < n && y + 1 < n && grid[x + 1][y + 1] == 0 && grid[x][y + 1] == 0 {
                    res.push((x, y, true));
                }

            }
            res
        }
        let n = grid.len();
        let state = (0, 0, true);
        let mut queue = VecDeque::from([(0, state)]);
        let mut visited = HashSet::from([state]);
        while let Some((cost, state)) = queue.pop_front() {
            if state == (n - 1, n - 2, true) {
                return cost;
            }
            for next_state in succesors(&grid, n, state).into_iter(){
                if ! visited.contains(&next_state) {
                    visited.insert(next_state);
                    queue.push_back((cost + 1, next_state));
                }
            }
        }
        -1
    }
}

#[test]
fn test_1() {
    let grid = vec![
        vec![0,0,0,0,0,1],
        vec![1,1,0,0,1,0],
        vec![0,0,0,0,1,1],
        vec![0,0,1,0,1,0],
        vec![0,1,1,0,0,0],
        vec![0,1,1,0,0,0]
        ];
    assert_eq!(11, Solution::minimum_moves(grid));
}

#[test]
fn test_2() {
    let grid = vec![
        vec![0,0,1,1,1,1],
        vec![0,0,0,0,1,1],
        vec![1,1,0,0,0,1],
        vec![1,1,1,0,0,1],
        vec![1,1,1,0,0,1],
        vec![1,1,1,0,0,0]
        ];
    assert_eq!(9, Solution::minimum_moves(grid));
}
